### PR TITLE
Fix mvnormal

### LINKED ISSUE #2023: Positive definite issue in MvNormal during Cholesky decomposition
Similar issue has been report also in #2015.
While I love the new cholesky decomposed covariance matrix in MvNormal, setting it as [a default](https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/multivariate.py#L70) is numerical quite unstable:
```python
        if cov is not None:
            self.chol_cov = tt.slinalg.cholesky(tt.as_tensor_variable(cov))
        elif tau is not None:
            self.chol_tau = tt.slinalg.cholesky(tt.as_tensor_variable(tau))
        else:
            self.chol_cov = tt.as_tensor_variable(chol)
```
I have a few simulation parameterized on `cov` used to work fine, they now returns error during the calling of `tt.slinalg.cholesky` quite often. 

Maybe we should use QR decomposition and reconstruct the cholesky factor? It should be more stable.

### PR DESCRIPTION
**Pull request by aseyboldt**
This should fix #2023 and #2019 and  #2015.

In #2023 we said we should go back to using a general purpose matrix solver instead of using the cholesky decomposition for now. But on second thought this doesn't seem right. With that we would have no way to tell if the covariance matrix is in fact positive definite. The old implementation did not check and just silently used the log(abs(det)) via the singular values. For ill conditioned and indefinite matrices it just returned wrong results.

I worked around the immediate problem using a custom cholesky op, that does not throw an exception, but instead returns nan. This leads to some more trouble in `solve_triangular`, but I think I got around that, too.

Maybe it would be better to fix this in the step methods instead. I can't see why they wouldn't be able to handle exceptions coming from the logp or dlogp computations. NUTS could handle those as divergent transitions and print the first couple of error messages in a trace as warnings. That would also be easier on the users, as those might point to problems in the model. We would need to go though all step methods, advi and all posterior sampling and check that they all get that right.

Todo
- [x] Test the examples (I'll do that tomorrow)
- [x] Figure out if we can improve the gradient computation. I think there is potential to do that quite a bit faster.
- [ ] Update lkj notebook
- [ ] Long term: Wait for pivoted cholesky in scipy.

### CODE REVIEW

File: pymc3/math.py [hunk 1]
```diff
@@ -12,9 +12,92 @@
...
 # pylint: enable=unused-import
 
 
+class Cholesky(Op):
+    """
+    Return a triangular matrix square root of positive semi-definite `x`.
+
+    This is a copy of the cholesky op in theano, that doesn't throw an
+    error if the matrix is not positive definite, but instead returns
+    nan.
```
Comment (twiecki) [inline pymc3/math.py]: @nouiz is that something you would consider to upstream theano?

**Comment by twiecki** [comment]
I agree that the step methods could check that. would it be a nan check? could do this in the base class potentially.

File: pymc3/math.py (hunk 1 above)
Comment (nouiz) [reply pymc3/math.py]: In fact, this is mostly a copy of an existing Theano op in tensor/slinalg.py I think the difference can be moved upstream.

Comment (aseyboldt) [reply pymc3/math.py]: I'll send a pull request.

**Comment by aseyboldt** [comment]
Not only a nan check. NUTS at least can already deal with nans. I think the step methods should catch some of the exceptions that can be produced when a theano function is evaluated. I think they should handle at least `scipy.linalg.LinalgError`, and treat them like a nan logp.
We might also want to  change `dist_math.bound` to optionally raise an exception instead of returning -inf. The samplers could treat that the same way as a -inf, but we would get a message telling the user why samples get rejected. Stan prints messages like that, and in my experience they make it much easier to debug an invalid model – the message says which parameter in which sampling statement had an invalid value. That would probably slow things down a bit, but we could make it optional (maybe through the model context). While we are at that, it would also be nice to have a way to disable the bounds checks altogether. That saves a few cycles and makes it easier for theano to optimize the graph.

**Comment by twiecki** [comment]
Regarding the bounds: completely agree, these should be largely impossible to hit with the transforms to begin with, so only add graph complexity. You propose to replace it with an on-call check that raises an exception instead of adding to the graph?

**Comment by aseyboldt** [comment]
I'd like something like this behaviour:
```python
with pm.Model() as model:
    sd = ...  # something strange that might be negative
    pm.Normal('mu', mu=0, sd=sd)
    pm.sample(...)
# output
Warning: Illegal parameter for variable mu: sd=-5. Rejecting sample.
Warning: Illegal parameter for variable mu: sd=-5. Rejecting sample.
Warning: Illegal parameter for variable mu: sd=-5. Rejecting sample.
Warning: Maximum number of rejection warnings reached. Silencing rejection warnings.
```
```python
with pm.Model(bounds_checks=False) as model:
    sd = ...  # something valid
    pm.Normal('mu', mu=0, sd=sd)
    pm.sample(...)
```
-> No warnings, but hopefully faster sampling.

**Comment by twiecki** [comment]
That looks good.

**Comment by aseyboldt** [comment]
I submitted the Cholesky op to Theano. I guess we should still keep it around in pymc3 until this is merged and released.

**Comment by twiecki** [comment]
@aseyboldt Awesome, and agreed.

**Comment by junpenglao** [comment]
@aseyboldt I know you are working on this so I wont open another issue:
In the `LKJCholeskyCov` [docstring](https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/multivariate.py#L603), you need to provide the `shape`: 
```vals = pm.MvNormal('vals', mu=np.zeros(10), chol=chol, shape=10)```

**Comment by aseyboldt** [comment]
On top of fixing the posdef issues, I have an implementation that takes a shortcut for the gradient, which leads to a ~2x speed increase for a leapfrog step. However, I ran into a theano bug: Theano/Theano#5908. This is fixed in master, but until it is released we can't use the faster implementation. I still included the code with tests in this PR, switching to the faster version will be very easy when the fix ends up in a theano release.
If the tests pass I think this is ready to merge.

**Comment by kyleabeauchamp** [comment]
FYI I think a few of the GPU fixes also depend on Theano master fixes.

**Comment by aseyboldt** [comment]
@kyleabeauchamp But it works on a CPU with the latest release, right?
If we use the latest released theano version for the faster implementation that breaks hessian computations.

**Comment by kyleabeauchamp** [comment]
AFAIK CPU still works on latest release

**Comment by twiecki** [comment]
@aseyboldt Is this ready to merge?

**Comment by aseyboldt** [comment]
yes.

File: pymc3/distributions/dist_math.py [hunk 2]
```diff
@@ -213,3 +216,149 @@ def logdet(m):
...
+
+class Cholesky(theano.Op):
+    """
+    Return a triangular matrix square root of positive semi-definite `x`.
+
+    This is a copy of the cholesky op in theano, that doesn't throw an
+    error if the matrix is not positive definite, but instead returns
+    nan.
+
+    L = cholesky(X, lower=True) implies dot(L, L.T) == X.
```
Comment (twiecki) [inline pymc3/distributions/dist_math.py]: maybe also add a note that this can be removed once a new theano version comes out.

Comment (aseyboldt) [reply pymc3/distributions/dist_math.py]: done.

**Comment by ilayn** [comment]
@twiecki I was going through the LDL factorization again to be (hopefully) available in version 1.1 of SciPy with https://github.com/scipy/scipy/pull/7941 regarding your last item above. Sorry for the delay, the LAPACK version update discussion took longer than we anticipated. Hence we decided to move forward with whatever we have in LAPACK 3.4.0. 

But from the discussion above, it seems that you are actually after a pivoted Cholesky (uses ?PSTRF routines) rather than LDL factorization (uses ?(HE/SY)TRF routines). Am I right? 

Then I can add the wrappers for low level calls which admittedly require more careful calls. They also don't provide a guarantee for definiteness in LAPACK routines.

**Pull request merged**